www/www.reactos.org/roslogin/admin/Mattermost.php

```diff
@@ -0,0 +1,228 @@
+			{
+				/* Request was successful, but response is not recognized */
+				$this->_last_error = 'Cannot connect to Mattermost';
+				return null;
+			}
+			return $json['id'];
+		}
+
+		/**
+		 * Returns user sessions
+		 */
+		public function getSessionsById($user_id)
+		{
+			$user_id = urlencode($user_id);
+			$json = $this->_ch->get("{$this->_url}/api/v4/users/{$user_id}/sessions");
+			if (!$json)
+			{
+				$this->_last_error = $this->_ch->getLastError();
+				return null;
+			}
+			$json = json_decode($json, true);
+			if (Mattermost::isError($json))
+			{
+				$this->_last_error = $json['message'];
+				return null;
+			}
+			return $json;
+		}
+
+		/**
+		 * Returns user audits
+		 */
+		public function getAuditsById($user_id)
+		{
+			$user_id = urlencode($user_id);
+			$json = $this->_ch->get("{$this->_url}/api/v4/users/{$user_id}/audits");
+			if (!$json)
+			{
+				$this->_last_error = $this->_ch->getLastError();
+				return null;
+			}
+			$json = json_decode($json, true);
+			if (Mattermost::isError($json))
+			{
+				$this->_last_error = $json['message'];
+				return null;
+			}
+			return $json;
+		}
+
+		/**
+		 * Revokes user session
+		 */
+		public function revokeSession($user_id, $session_id)
+		{
+			$user_id = urlencode($user_id);
+			$json = $this->_ch->post(
+				"{$this->_url}/api/v4/users/{$user_id}/sessions/revoke",
+				"application/json",
+				json_encode(array('session_id' => $session_id)));
+			if (!$json)
+			{
+				$this->_last_error = $this->_ch->getLastError();
+				return null;
+			}
+			$json = json_decode($json, true);
+			if (Mattermost::isError($json))
+			{
+				$this->_last_error = $json['message'];
+				return null;
+			}
+			return $json;
+		}
+
+		/**
+		 * Revokes all user sessions
+		 */
+		public function revokeAllSessions($user_id)
+		{
+			$user_id = urlencode($user_id);
+			$json = $this->_ch->post(
+				"{$this->_url}/api/v4/users/{$user_id}/sessions/revoke/all",
+				"application/json",
+				"{}");
+			if (!$json)
+			{
+				$this->_last_error = $this->_ch->getLastError();
+				return null;
+			}
+			$json = json_decode($json, true);
+			if (Mattermost::isError($json))
+			{
+				$this->_last_error = $json['message'];
+				return null;
+			}
+			return $json;
+		}
+
+		/**
+		 * Returns all IP addresses associated with a session
+		 * using audit information
+		 */
+		public function getSessionAddresses($audits, $session_id)
+		{
+			$ips = array();
+			for ($i = 0; $i < count($audits); $i++)
+			{
+				$e = $audits[$i];
+				if ($e['session_id'] === $session_id)
+				{
+					$ip = $e['ip_address'];
+					if (!in_array($ip, $ips, true))
+						$ips[] = $ip;
+				}
+			}
+			return $ips;
+		}
+
+		/**
+		 * Writes error message to disk to display it later
+		 */
+		public function rememberError()
+		{
+			$error = $this->getLastError();
+			$filename = Mattermost::getTempFilename();
+			file_put_contents($filename, $error);
+			return $error;
+		}
```

Comment:
I don't like RosLogin writing anything to disk and thus becoming stateful.
Possible expected errors (like "User has no active Mattermost sessions") should be handled in UserPage.php. For any other unexpected error, the action scripts should just throw a new `RuntimeException` like the others are already doing.
Even if that doesn't look nice, there is no reason to pipe the exact error back to UserPage.php using a temp file.